Kraken: additional attributes for realtime trip creation

pbougue opened the pull request:
:mag: Please review by commit with message (cleaning and preparatory work is separated).

 :warning: Major bump of version (bina required)

Handle new attributes (when filled) from NTFSRT-TripUpdate when creating a realtime VJ (instead of using default values).
Requirement : same behavior than before when values are empty (existing tests should not change).

TODO:
- [x] Tests will be done in separate commits (hopefully not changing the current code :crossed_fingers:)

Supersedes #4058 
JIRA: https://navitia.atlassian.net/browse/NAV-1926

Commit:
Side-cleaning: rename and improve comment for headsign_handler

To have all matching code (after functionnal changes were made).

Commit:
PT-Data: extract getters from get_or_create

To be able to use getter alone in next commit

Commit:
Kraken: handle additional attributes for realtime VJ-creation

/!\ bump of data_version (requires binarization)

Tests to come in a separate commit.

pbougue commented on source/type/message.cpp:
```diff
@@ -118,7 +118,8 @@ SERIALIZABLE(ApplicationPattern)
 void Impact::serialize(Archive& ar, const unsigned int /*unused*/) {
     ar& uri& company_id& physical_mode_id& headsign& created_at& updated_at& application_periods& severity&
-        _informed_entities& messages& disruption& aux_info& application_patterns;
+        _informed_entities& messages& disruption& aux_info& application_patterns& trip_short_name& dataset_id&
+            network_id& commercial_mode_id& line_id& route_id;
```
Required for data cloning (RT or base-schedule reload) : https://github.com/hove-io/navitia/blame/714f409886302e6b38a195c4ee4e9db66eb0c199/source/type/data.cpp#L874-L883

xlqian replied:
it's probably possible to load the old data without re-bina.... have you tried to load the `data.nav` with this PR?

pbougue replied:
After discussion, let's downgrade back to old data_version.

Commit:
RT trip creation: add kraken test on new attributes

Happy-path only

kadhikari commented on source/kraken/apply_disruption.cpp:
```diff
@@ -200,21 +200,66 @@ type::ValidityPattern compute_base_disrupted_vp(const std::vector<boost::posix_t
+        route_name = "Additional service";
+        direction_type = "outbound";
+    } else {
+        route_uri = "route:" + sa_depart->uri + "_" + sa_arrival->uri;
```
```suggestion
        route_uri = "route:" + line->uri
```

kadhikari replied:
OK for me as i found:
Requirement : same behavior than before when values are empty (existing tests should not change).

kadhikari commented on source/kraken/apply_disruption.cpp:
```diff
@@ -200,21 +200,66 @@ type::ValidityPattern compute_base_disrupted_vp(const std::vector<boost::posix_t
+    } else {
+        route_uri = "route:" + sa_depart->uri + "_" + sa_arrival->uri;
+        route_name = sa_depart->name + " - " + sa_arrival->name;
+        direction_type = "forward";
```
```suggestion
        direction_type = "outbound";
```

kadhikari replied:
Be sure of name, direction type .. as explain in the doc

kadhikari replied:
OK for me as i found: 
Requirement : same behavior than before when values are empty (existing tests should not change).

kadhikari commented on source/kraken/apply_disruption.cpp:
```diff
@@ -424,21 +472,35 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                }
+                if (dataset != nullptr) {
+                    vj->dataset = dataset;
+                    vj->dataset->vehiclejourney_list.insert(vj);
```
Minor: Can this block be moved above and be include in 
if (dataset == nullptr && !pt_data.datasets.empty() && pt_data.datasets[0]) {

pbougue replied:
Letting it here allows checking that dataset is not `nullptr` after previous block (in case `pt_data.datasets[0]`is null).
In this case, this line would crash.
So really unlikely, but protecting a crash was the goal here.

kadhikari replied:
:+1:

Commit:
RT trip creation: fix reject on wrong attributes

Add kraken test rejecting disruption.
Correction of code to correctly reject those cases.

Commit:
Add some tests on new trip with additional attributes

Commit:
Little formatting

Commit:
Little adjustment on test

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -3405,6 +3604,165 @@ BOOST_FIXTURE_TEST_CASE(company_id_doesnt_exist_in_new_trip, AddTripDataset) {
+        return pb_creator.get_response();
+    };
+
+    // If the company id doesn't exist inside the data with ADDED type, we reject the new trip
```
```suggestion
    // If the line id doesn't exist inside the data with ADDED type, we reject the new trip
```

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -3405,6 +3604,165 @@ BOOST_FIXTURE_TEST_CASE(company_id_doesnt_exist_in_new_trip, AddTripDataset) {
+        return pb_creator.get_response();
+    };
+
+    // If the company id doesn't exist inside the data with ADDED type, we reject the new trip
```
```suggestion
    // If the dataset id doesn't exist inside the data with ADDED type, we reject the new trip
```

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -3405,6 +3604,165 @@ BOOST_FIXTURE_TEST_CASE(company_id_doesnt_exist_in_new_trip, AddTripDataset) {
+        return pb_creator.get_response();
+    };
+
+    // If the company id doesn't exist inside the data with ADDED type, we reject the new trip
```
```suggestion
    // If the network id doesn't exist inside the data with ADDED type, we reject the new trip
```

kadhikari commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -3405,6 +3604,165 @@ BOOST_FIXTURE_TEST_CASE(company_id_doesnt_exist_in_new_trip, AddTripDataset) {
+        return pb_creator.get_response();
+    };
+
+    // If the company id doesn't exist inside the data with ADDED type, we reject the new trip
```
```suggestion
    // If the commercial_mode id doesn't exist inside the data with ADDED type, we reject the new trip
```

Commit:
Bump chaos-proto to master

Also little update on comments after review

xlqian commented on source/kraken/realtime.cpp:
```diff
@@ -320,23 +320,61 @@ static bool is_handleable(const transit_realtime::TripUpdate& trip_update,
-            return false;
+        if (is_added_trip(trip_update)) {
+            // check if company id exists
+            if (trip_update.trip().HasExtension(kirin::company_id)
```
minor: it'll be much more readable you can put all these codes in functions... like `check_company_id()` `check_physical_mode_id()`

besides, these blocs of codes look quite similar...

pbougue replied:
I'm not willing to dig more on templating.
`kirin::company_id`'s type is `google::protobuf::internal::ExtensionIdentifier< ::transit_realtime::TripDescriptor, ::google::protobuf::internal::StringTypeTraits, 9, false>` :sweat_smile: and 4 others have the same type, but not physical_mode_id.
But I agree on readability for that, some intermediate steps may be done, I won't dig, but it would be legitimate IMO.

Commit:
Back to old data_version

As no impact exists in nav files, it should be OK to keep last version.
The new way of serializing/deserializing will be used on clone.
Therefore, expecting all OK.
